crates/sui-types/src/messages.rs

```diff
@@ -1062,10 +1090,149 @@ impl PartialEq for SignedTransaction {
+///
+/// DO NOT USE in networked APIs.
+///
+/// Because it is used very sparingly, it can be audited easily: Use rust-analyzer, or run:
+///
+///      $ git grep -E 'TrustedTransactionEnvelope|TrustedCertificate'
+///
+/// And verify that none of the uses appear in any network APIs.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct TrustedTransactionEnvelope<T>(TransactionEnvelope<T>);
+
+impl<T> TrustedTransactionEnvelope<T> {
+    pub fn into_inner(self) -> TransactionEnvelope<T> {
+        self.0
+    }
+}
+
+impl<T> Debug for TrustedTransactionEnvelope<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.0)
+    }
+}
+
+// An empty marker struct that can't be serialized.
+#[derive(Clone)]
+struct NoSer;
+
+/// VerifiedTransactionEnvelope models a TransactionEnvelope that no longer requires signature
+/// verification. It can only be constructed in 3 ways:
+///
+/// - TransactionEnvelope::verify() - verify the signatures of a signed transaction, to convert a
+///   TransactionEnvelope<T> into a VerifiedTransactionEnvelope<T>
+/// - From<TrustedTransactionEnvelope<T>> - used when reading pre-verified messages from the
+///   database.
+/// - VerifiedTransactionEnvelope::new_unchecked() - use very carefully, and very sparingly, so
+///   that all uses can be audited easily.
+#[derive(Clone)]
+pub struct VerifiedTransactionEnvelope<T>(TrustedTransactionEnvelope<T>, NoSer);
+
+// Never remove this assert!
+static_assertions::assert_not_impl_any!(VerifiedTransactionEnvelope: Serialize, DeserializeOwned);
+
+impl<T> Debug for VerifiedTransactionEnvelope<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self.0 .0)
+    }
+}
+
+impl<T> VerifiedTransactionEnvelope<T> {
+    fn new_from_verified(inner: TransactionEnvelope<T>) -> Self {
+        Self(TrustedTransactionEnvelope(inner), NoSer)
+    }
+
+    /// There are some situations (e.g. fragment verification) where its very awkward and/or
+    /// inefficient to obtain verified certificates from calling CertifiedTransaction::verify()
+    /// Use this carefully.
+    pub fn new_unchecked(inner: TransactionEnvelope<T>) -> Self {
+        Self(TrustedTransactionEnvelope(inner), NoSer)
+    }
+
+    pub fn into_inner(self) -> TransactionEnvelope<T> {
+        self.0 .0
+    }
+
+    /// Use this when you need to serialize a verified envelope.
+    /// This should generally only be used for database writes.
+    /// ***never use over the network!***
+    pub fn serializable_ref(&self) -> &TrustedTransactionEnvelope<T> {
+        &self.0
+    }
+
+    /// Use this when you need to serialize a verified envelope.
+    /// This should generally only be used for database writes.
+    /// ***never use over the network!***
+    pub fn serializable(self) -> TrustedTransactionEnvelope<T> {
+        self.0
+    }
```

Comment:
My one concern is around how easy it is to call this. One other way to do this would be to not have the `Trusted*` types at all and instead always have the DBs serialize the unverified types, but the apis which we call to insert and read the DB only expose the Verified types. That would require a few more `new_unchecked` calls on the read side of the DBs.

Reply:
I thought about doing this, but I think the Trusted* approach is better, since you only need to audit where Trusted* is used, and it should only be used in database table definitions, which inherently shouldn't proliferate. Whereas the new_unchecked approach means that you need to call new_unchecked anytime you do a db read, and reads can proliferate due to the various multi-read and iterator patterns we use.

Note that calling serializable/serializable_ref isn't actually dangerous. It only becomes dangerous the minute you include it in a type that is sent over the network. But since, at least for now, Trusted* should never appear _anywhere_ except database tables, its very easy to audit.